enhancements/microshift/microshift-router-configuration.md

```diff
@@ -0,0 +1,441 @@
+upon starting. This includes not just the pod, but all the associated resources
+that come along with it: namespace, services, configmaps, etc.
+
+When setting the option to `Disabled`, the next MicroShift restart should delete
+all the router related resources, including APIs and controllers.
+
+Setting the option to `Enabled` will deploy the router, while `Disabled` will disable
+it. Default will be `Enabled`.
+
+### Listening ports
+The following configuration is proposed:
+```yaml
+ingress:
+  ports:
+    http: <int> # Defaults to 80.
+    https: <int> # Defaults to 443.
+```
+
+MicroShift does not own the host, which means there might be other
+applications running alongside and some ports may already be opened.
+MicroShift must be able to accommodate such situations and be ready to listen
+in other ports, hence the configuration options.
+
+In order to allow this configuration and many other advantages, the
+router service shall be changed so that it is exposed using `LoadBalancer`
+type instead of using host ports. See the Design Details section for more
+information.
+
+#### Firewalling ports
+Using `LoadBalancer` service type prevents the usage of firewalld to block
+access to ports. Since this is a feature that users may require for auditing,
+an alternative is provided.
+
+The `NetworkPolicy` resource fits the needs for blocking traffic to services
+and pods. Policies can also be configured to allow specific IP addresses and
+provide a high degree of customization for network traffic. They use the CNI
+instead of host level settings, providing greater flexibility and being self
+contained within MicroShift.
+
+Users in need of blocking traffic shall use the `NetworkPolicy` resource. A
+section in the documentation will be added to describe how to do this.
+
+### Configure IPs where the router listens
+The following configuration is proposed:
+```yaml
+ingress:
+  expose:
+    - <IP address/hostname> # Defaults to the node's hostname.
+```
+
+As we have seen in previous sections, the use of `LoadBalancer` makes ovnk
+configure iptables rules to expose the service outside of the cluster.
+Building on [LoadBalancer controller enhancement](loadbalancer-service-support.md),
+the service is exposed using the node IP and configured ports.
+
+All the IPs where the router is exposed need to be part of the `LoadBalancer`
+service status definition. In order to provide more flexibility, the
+configuration options are able to take IP addresses and/or hostnames.
+
+MicroShift will use the hostname as a default value for this. The router will
+only listen on the primary IP for the host unless configured to do otherwise.
+This is a sensible default that covers basic networking setups. Any advanced
+configuration, such as multiple interfaces, VLANs, etc. will need user's
+configuration.
+The hostnames will be turned into the IP addresses automatically by
+MicroShift, which will also watch for IP changes by doing periodic queries to
+the DNS.
```

Comment:
I'm still not clear about why we would want to take hostnames. It sounds like it is going to be more work to implement and require yet another periodic task to consume CPU. What does using hostnames give us that using IPs alone does not?

Reply:
For example, relying on DNS or DHCP for the host. I am thinking of cases where you would have another NIC (not the primary one) where you would rely on DNS from external sources and changing its IP would not trigger a MicroShift restart.